### PR TITLE
Add new .NET Core 3.1 version of the Lambda test tool

### PR DESCRIPTION
**Pull request by normj**
Add support for .NET Core 3.1 with the .NET Lambda Test Tool. This PR looks much larger then it really is because of massive file rearrangements. The PR was created by first creating a new branch with the files removed and then a second branch was created from the first with the new files added. This was necessary due to the 1000+ files that were previously included in the repo for bootstrap and font awesome packages. Now just the min versions are used.

The Lambda Test Tool must match the version of .NET Core that developers Lambda function is written in. Currently we ship the NuGet package **Amazon.Lambda.TestTool-2.1** which has a cli command called **dotnet-lambda-test-tool-2.1**. With the 3.1 support we will ship an additional tool called **Amazon.Lambda.TestTool-3.1** containing **dotnet-lambda-test-tool-3.1**.

The project was refactored to put all of the common logic about selected Lambda function into the **Amazon.Lambda.TestTool** project. Then separate ASP.NET Core applications for 2.1 and 3.1 were created. The 2.1 is the same ASP.NET Core that we currently ship for 2.1 support. The 3.1 is a new application using Blazor Server Side technology.

image

Besides supporting .NET Core 3.1 the new features/fixes added are:
* The selected Lambda code is now loaded in a separate AssemblyLoadContext to avoid assembly load collisions that current users are having issues. This logic can be found at `Amazon.Lambda.TestTool.Runtime.LambdaAssemblyLoadContext`.
* The test tool can now be run with the `--no-ui` switch to avoid using the web interface to select the code and use command line arguments to identify the function. This logic can be found at `Amazon.Lambda.TestTool.TestToolStartup.ExecuteWithNoUi()`.
* A new documentation page was added to the 3.1 version.
* The settings folder will now only be created during write operations.

The tests were broken out to separate projects for 2.1 and 3.1 and used a shared source project. This was done instead of multi targeting because mac builds wouldn't work with the conditional project reference includes.

Areas to focus for PR:
* All of the **Amazon.Lambda.TestTool.WebTester31** project.
* The **TestToolStartup** class which contains the logic for the new --no-ui mode.
* The **LambdaAssemblyLoadContext** class containing the custom AssemblyLoadContext

The **Amazon.Lambda.TestTool.WebTester21** is the exact same ASP.NET Code that ships today for the .NET Core 2.1 version.

### CODE REVIEW

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/TestToolStartup.cs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+                    {
+                        Console.WriteLine("Press any key to exit");
+                        Console.ReadKey();
+                    }
+                    System.Environment.Exit(-2);
+                }
+            }
+        }
+        
+        public static void ExecuteWithNoUi(LocalLambdaOptions localLambdaOptions, CommandLineOptions commandOptions, string lambdaAssemblyDirectory, RunConfiguration runConfiguration)
```
Comment (awschristou) [inline Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/TestToolStartup.cs]: Suggest splitting this function into smaller functions

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/FunctionPickerComponent.razor [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    void UpdateConfigFileChange()
+    {
+        var configInfo = LambdaDefaultsConfigFileParser.LoadFromFile(this.ConfigFile);
+        AvailableFunctions = this.LambdaOptions.LambdaRuntime.LoadLambdaFunctions(configInfo.FunctionInfos);
+        if(AvailableFunctions.Count > 0)
+        {
+            this.FunctionHandler = this.AvailableFunctions[0].FunctionInfo.Handler;
+        }
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/FunctionPickerComponent.razor]: Done

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/Index.razor [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        if(SampleRequestManager.TryDetermineSampleRequestName(SelectedSampleRequestName, out var sampleRequestName))
+        {
+            parameters.Add(SaveRequestDialog.PARAMETER_NAME_CURRENT_NAME, sampleRequestName);
+        }
+
+        Modal.OnClose += OnSaveRequestClosed;
+        Modal.Show<SaveRequestDialog>("Save Request", parameters);
+    }
+
+    void OnSaveRequestClosed(ModalResult modalResult)
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/Index.razor]: Good catch. Fixed.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/SaveRequestDialog.razor [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿@using Amazon.Lambda.TestTool.SampleRequests;
+@inject LocalLambdaOptions LambdaOptions
+@inject IModalService ModalService
+
+<div class="simple-form" style="min-width: 400px">
+    <div class="form-group" for="save-request-name">
+        <label class="control-label">Please enter a request name: </label>
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/SaveRequestDialog.razor]: Removed. I also removed it from the .NET Core 2.1 version.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/_Host.cshtml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+﻿@page "/"
+@namespace Amazon.Lambda.TestTool.WebTester31.Pages
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
+@{
+    Layout = null;
+
+    var overlayMessage = $"{Constants.PRODUCT_NAME} has paused. This can happen while at a breakpoint. This can happen for the following reasons. <ul><li>The test tool will resume when execution has continued in the debugger.</li><li>If the debugger has been stopped this window can be closed.</li></ul>";
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/_Host.cshtml]: Removed. It was left over from an earlier attempt to get the messaging up.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/_Host.cshtml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+<body>
+
+    <div id="components-reconnect-modal" class="my-reconnect-modal components-reconnect-hide">
+        <div class="show failed refused">
+            <div class="center-text">
+                <p>
+                    @Constants.PRODUCT_NAME has paused. This can happen for the following reasons.
+                    <ul>
+                        <li>The attached debugger is currently on a breakpoint. The test tool will resume when execution has continued in the debugger.</li>
+                        <li>The debugger or test tool process has stopped in which case this window can be closed.</li>
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Pages/_Host.cshtml]: Replaced with your suggested verbiage.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/CommandLineOptions.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            Console.WriteLine();
+            
+            Console.WriteLine("These options are valid in the no web interface mode.");
+            Console.WriteLine();
+            Console.WriteLine("\t--no-ui                               Disable launching the web interface and immediately execute the Lambda code.");
+            Console.WriteLine("\t--profile <profile-name>              Set the AWS credentials profile to provide credentials to the Lambda code.");
+            Console.WriteLine("\t                                      If not set the profile from the config file will be used.");
+            Console.WriteLine("\t--region <region-name>                Set the AWS region to as the default region for the Lambda code being executed.");
+            Console.WriteLine("\t                                      If not set the region from the config file will be used.");
+            Console.WriteLine("\t--config-file <file-name>             The config file to read for Lambda settings. If not set than aws-lambda-tools-defaults.json");
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/CommandLineOptions.cs]: Fixed.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/Resources/SampleRequests/CodeCommit.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
+{
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/Resources/SampleRequests/CodeCommit.json]: They were already in the repo. I could look into automatically pulling them as a build step in the future but I don't want to take on that work now.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Shared/MainLayout.razor [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+<div class="sidebar">
+    <NavMenu />
+</div>
+
+<div class="main">
+    <div class="content px-4">
+        @Body
+        <hr />
+        <footer>
+            <p>@Constants.PRODUCT_NAME (@Utils.DetermineToolVersion()) is an open source project. For issues or feedback visit the <a href="https://github.com/aws/aws-lambda-dotnet/tree/master/Tools/LambdaTestTool">AWS Lambda for .NET Core GitHub repository.</a></p>
```
[the remark this answers is not in the document]
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.WebTester31/Shared/MainLayout.razor]: Done for the .NET Core 3.1 version putting the links in the Constants class.

File: Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/TestToolStartup.cs (hunk 1 above)
Comment (normj) [reply Tools/LambdaTestTool/src/Amazon.Lambda.TestTool/TestToolStartup.cs]: Yeah that method got a bit carried away. Its broken up now.

**Comment by normj** [comment]
Merged into dev-netcore31 branch

**Pull request closed without merging**